### PR TITLE
xTransfer update

### PR DESCRIPTION
**Pull request by tolak**
- Deduct fee when forwarding assets to solo chain.
- Pay asset itself as fee rather than PHA when transfering non-native assets to solo chains.
- Extra events added in xcm-transfer pallet.
- More information on chainbridge registry struct

### CODE REVIEW

File: pallets/xtransfer/src/xcm/xcm_helper.rs [hunk 1]
```diff
@@ -291,11 +358,14 @@ pub mod xcm_helper {
 				&what,
 				&who,
 			);
-			if NativeChecker::is_native_asset(what) {
-				NativeAdapter::withdraw_asset(what, who)
+			let assets = if NativeChecker::is_native_asset(what) {
+				NativeAdapter::withdraw_asset(what, who).map_err(|e| return e)?
```
Comment (h4x3rotab) [inline pallets/xtransfer/src/xcm/xcm_helper.rs:362]:
```suggestion
			let assets = if NativeChecker::is_native_asset(what) {
				NativeAdapter::withdraw_asset(what, who)?
```

File: pallets/xtransfer/src/xcm/xcm_helper.rs [hunk 2]
```diff
@@ -291,11 +358,14 @@ pub mod xcm_helper {
...
 			} else {
-				AssetsAdapter::withdraw_asset(what, who)
-			}
+				AssetsAdapter::withdraw_asset(what, who).map_err(|e| return e)?
```
Comment (h4x3rotab) [inline pallets/xtransfer/src/xcm/xcm_helper.rs:364]:
```suggestion
				AssetsAdapter::withdraw_asset(what, who)?
```

File: pallets/xtransfer/src/xcm/xcm_helper.rs [hunk 3]
```diff
@@ -291,11 +358,14 @@ pub mod xcm_helper {
...
+			};
+			WithdrawHandler::on_withdrawn(what.clone(), who.clone())
```
Comment (h4x3rotab) [inline pallets/xtransfer/src/xcm/xcm_helper.rs:366]:
```suggestion
			WithdrawHandler::on_withdrawn(what, who)
```

File: pallets/xtransfer/src/xcm/xcm_helper.rs [hunk 4]
```diff
@@ -264,21 +329,23 @@ pub mod xcm_helper {
...
-						NativeAdapter::deposit_asset(what, who)
+						NativeAdapter::deposit_asset(what, who).map_err(|e| return e)?;
 					} else {
-						AssetsAdapter::deposit_asset(what, who)
+						AssetsAdapter::deposit_asset(what, who).map_err(|e| return e)?;
 					}
 				}
 			}
+
+			DepositHandler::on_deposited(what.clone(), who.clone())
```
Comment (h4x3rotab) [inline pallets/xtransfer/src/xcm/xcm_helper.rs:346]: ditto

File: pallets/xtransfer/src/xcm/xcm_helper.rs [hunk 5]
```diff
@@ -264,21 +329,23 @@ pub mod xcm_helper {
...
 					)
 					.map_err(|e| XcmError::FailedToTransactAsset(e.into()))?;
-
-					Ok(())
 				}
 				// Try handle it with transfer adapter
 				_ => {
 					if NativeChecker::is_native_asset(what) {
-						NativeAdapter::deposit_asset(what, who)
+						NativeAdapter::deposit_asset(what, who).map_err(|e| return e)?;
```
Comment (h4x3rotab) [inline pallets/xtransfer/src/xcm/xcm_helper.rs:339]: ditto

File: pallets/xtransfer/src/bridge_transfer/mod.rs [hunk 6]
```diff
@@ -480,4 +505,45 @@ pub mod pallet {
...
+									/ T::NativeExecutionPrice::get(),
+							)
+						} else {
+							fee_in_asset = None
+						}
+						fee_in_asset
+					}
+				}
+				_ => None,
+			};
```
Comment (h4x3rotab) [inline pallets/xtransfer/src/bridge_transfer/mod.rs:540]:
```suggestion
			match (&asset.id, &asset.fun) {
				(Concrete(asset_id), Fungible(amount)) => {
					let fee_estimated_in_pha =
						Self::estimate_fee_in_pha(chain_id, (*amount).into());
					if T::NativeChecker::is_native_asset(asset) {
						Some(fee_estimated_in_pha.into())
					} else {
						let fee_prices = T::ExecutionPriceInfo::get();
						if let Some(idx) = fee_prices.iter().position(|(fee_asset_id, _)| {
							fee_asset_id == &Concrete(asset_id.clone())
						}) {
							Some(
								fee_estimated_in_pha.into() * fee_prices[idx].1
									/ T::NativeExecutionPrice::get(),
							)
						} else {
							None
						}
					}
				}
				_ => None,
			};
```

File: pallets/xtransfer/src/assets_wrapper.rs [hunk 7]
```diff
@@ -280,6 +287,8 @@ pub mod pallet {
 			origin: OriginFor<T>,
 			asset_id: T::AssetId,
 			chain_id: u8,
+			is_mintable: bool,
+			memo: Box<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/xtransfer/src/assets_wrapper.rs]: Update to recommended spec

File: pallets/xtransfer/src/assets_wrapper.rs [hunk 8]
```diff
@@ -326,10 +346,15 @@ pub mod pallet {
...
-				}
+				let is_chainbridge = match item {
+					XBridge::ChainBridge {
+						chain_id: cid,
+						resource_id: rid,
+						..
+					} => cid == &chain_id && rid == &resource_id,
+					_ => false,
+				};
+				is_chainbridge
```
[the remark this answers is not in the document]
Comment (tolak) [reply pallets/xtransfer/src/assets_wrapper.rs]: done

File: pallets/xtransfer/src/xcm/xcm_helper.rs (hunk 1 above)
Comment (tolak) [reply pallets/xtransfer/src/xcm/xcm_helper.rs]: Done

File: pallets/xtransfer/src/xcm/xcm_helper.rs (hunk 2 above)
Comment (tolak) [reply pallets/xtransfer/src/xcm/xcm_helper.rs]: Done

File: pallets/xtransfer/src/xcm/xcm_helper.rs (hunk 3 above)
Comment (tolak) [reply pallets/xtransfer/src/xcm/xcm_helper.rs:366]: We can not share reference to `on_withdraw` here, `Event` arguments with a reference will need us give a life time parameter, not sure which one is more effective

File: pallets/xtransfer/src/xcm/xcm_helper.rs (hunk 5 above)
Comment (tolak) [reply pallets/xtransfer/src/xcm/xcm_helper.rs]: Done

File: pallets/xtransfer/src/bridge_transfer/mod.rs (hunk 6 above)
Comment (tolak) [reply pallets/xtransfer/src/bridge_transfer/mod.rs]: Done

File: pallets/xtransfer/src/xcm/xcm_helper.rs (hunk 3 above)
Comment (h4x3rotab) [reply pallets/xtransfer/src/xcm/xcm_helper.rs:366]: Oh, got it. It's already a reference.

**Review by h4x3rotab — APPROVED** [review]

**Pull request merged**